metric_learn/lmnn.py

```diff
@@ -101,233 +112,578 @@ class LMNN(MahalanobisMixin, TransformerMixin):
 
   Attributes
   ----------
+  components_ : `numpy.ndarray`, shape=(n_components, n_features)
+    The learned linear transformation ``L``.
+
+  n_neighbors_ : int
+      The provided ``n_neighbors`` is decreased if it is greater than or
+      equal to  min(number of elements in each class).
+
   n_iter_ : `int`
     The number of iterations the solver has run.
 
-  components_ : `numpy.ndarray`, shape=(n_components, n_features)
-    The learned linear transformation ``L``.
+  random_state_ : numpy.RandomState
+    Pseudo random number generator object used during initialization.
 
   Examples
   --------
 
   >>> import numpy as np
   >>> from metric_learn import LMNN
+  >>> from sklearn.neighbors import KNeighborsClassifier
   >>> from sklearn.datasets import load_iris
   >>> iris_data = load_iris()
   >>> X = iris_data['data']
   >>> Y = iris_data['target']
-  >>> lmnn = LMNN(k=5, learn_rate=1e-6)
-  >>> lmnn.fit(X, Y, verbose=False)
+  >>> lmnn = LMNN(n_neighbors=3)
+  >>> lmnn.fit(X, Y)
+  >>> knn = KNeighborsClassifier(n_neighbors=3)
+  >>> knn.fit(lmnn.transform(X), Y)
 
   References
   ----------
-  .. [1] K. Q. Weinberger, J. Blitzer, L. K. Saul. `Distance Metric
-         Learning for Large Margin Nearest Neighbor Classification
-         <http://papers.nips.cc/paper/2795-distance-metric\
-         -learning-for-large-margin-nearest-neighbor-classification>`_. NIPS
-         2005.
+  .. [1] Weinberger, Kilian Q., and Lawrence K. Saul.
+         "Distance Metric Learning for Large Margin Nearest Neighbor
+         Classification."
+         Journal of Machine Learning Research, Vol. 10, Feb. 2009,
+         pp. 207-244.
+         http://jmlr.csail.mit.edu/papers/volume10/weinberger09a/weinberger09a.pdf
+
+  .. [2] Wikipedia entry on Large Margin Nearest Neighbor
+         https://en.wikipedia.org/wiki/Large_margin_nearest_neighbor
+
   """
+  def __init__(self, n_neighbors=3, n_components=None, init='auto',
+               max_impostors=500_000, neighbors_params=None,
+               push_loss_weight=0.5, max_iter=50, tol=1e-5,
+               preprocessor=None, verbose=False, random_state=None):
 
-  def __init__(self, init='auto', k=3, min_iter=50, max_iter=1000,
-               learn_rate=1e-7, regularization=0.5, convergence_tol=0.001,
-               verbose=False, preprocessor=None,
-               n_components=None, random_state=None):
+    self.n_neighbors = n_neighbors
```

Comment:
Yes, I'll push these changes. I think `regularization` is not that clear, since one could consider either the push loss or the pull loss as the main data term. In my mind, it makes more sense to consider the pull loss as a regularization term. So I think it makes sense to make it explicit.

Reply:
Thanks for having added the deprecation warnings, yes I think this formulation is better indeed, more explicit